### PR TITLE
executor: check privilege before adding

### LINKED ISSUE #22703: grant Super/GRANT for user failed
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
create user `br`@`%` identified by 'br';
grant Super on test_cosven.* to 'br'@'%';
grant GRANT on test_cosven.* to 'br'@'%';
```

### 2. What did you expect to see? (Required)

no error

### 3. What did you see instead (Required)

```sql
mysql root@127.0.0.1:mysql> grant Super on test_cosven.* to 'br'@'%';
(1054, "Unknown column 'Super_priv' in 'field list'")
mysql root@127.0.0.1:mysql> grant SUPER on test_cosven.* to 'br'@'%';
(1054, "Unknown column 'Super_priv' in 'field list'")
mysql root@127.0.0.1:mysql> grant INSERT on test_cosven.* to 'br'@'%';
Query OK, 0 rows affected
Time: 0.035s
mysql root@127.0.0.1:mysql> grant ALTER on test_cosven.* to 'br'@'%';
Query OK, 0 rows affected
Time: 0.028s
mysql root@127.0.0.1:mysql> grant GRANT on test_cosven.* to 'br'@'%';
(1064, 'You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use line 1 column 14 near "on test_cosven.* to \'br\'@\'%\'" ')
```

### 4. What is your TiDB version? (Required)

```
Release Version: v5.0.0-rc\nEdition: Community\nGit Commit Hash: fd4437dde445012f5b75e5d4f336433cb076c1ee\nGit Branch: heads/refs/tags/v5.0.0-rc\nUTC Build Time: 2021-01-12 06:00:03\nGoVersion: go1.13\nRace Enabled: false\nTiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306\nCheck Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by xhebox**
### What problem does this PR solve?

Issue Number: close #22703 

Problem Summary: The original error is that we did not check if the priv to add is in the DB/TABLE/COLUMN list. So it becomes an anti-intuitive planner error. This PR fixed it by explicit checks.

The error message returned is different from mysql: error message of mysql is introduced in https://github.com/mysql/mysql-server/commit/0ed74a1a0c4f2b93cd3450dd3c971dfa21f31663#diff-2e64adb7f45663cd160f83990253588843690af33ddce7dd71a802ec6f86ad99R1750. At the time, there is only "DB GRANT" and "GLOBAL GRANT", so it is `incorrect usage of DB GRANT AND GLOBAL PRIVILEGES`. I've decided to return messages like `incorrect usage of XX GRANT AND NON-XX PRIVILEGES`.

`ErrIllegalGrantForTable` looks like that another mysql dev did not know the original `WRONG_USAGE` error, so he added another errcode. Maybe break the compatibility and use `WRONG_USAGE` for all error messages later.

It is a minor issue, so I am not going to cherry-pick.

It depends on https://github.com/pingcap/parser/pull/1201, which is already merged.

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- No release note

### CODE REVIEW

File: executor/revoke_test.go [hunk 1]
```diff
@@ -92,8 +92,16 @@ func (s *testSuite1) TestRevokeTableScope(c *C) {
 		c.Assert(rows, HasLen, 1)
 		row := rows[0]
 		c.Assert(row, HasLen, 1)
-		p := fmt.Sprintf("%v", row[0])
-		c.Assert(strings.Index(p, mysql.Priv2SetStr[v]), Equals, -1)
+
+		op := mysql.Priv2SetStr[v]
```
Comment (xhebox) [inline executor/revoke_test.go:96]: `string.Index("Create View,Drop", "Create")` will match.

File: executor/utils.go [hunk 2]
```diff
@@ -13,15 +13,23 @@
...
 
-import "strings"
+import (
+	"strings"
+)
 
-func setFromString(value string) []string {
+// SetFromString constructs a slice of strings from a comma separated string.
+// It is assumed that there is no duplicated entry. You could use addToSet to maintain this property.
+func SetFromString(value string) []string {
```
Comment (tiancaiamao) [inline executor/utils.go:22]: It's not used outside this package, why it's exported?

Comment (xhebox) [reply executor/utils.go:22]: Exported for tests.

Tests in executor is all named `_test`, thus tests are treated as a standalone package, which means they can not invoke internal/unexported functions. A [link](https://github.com/pingcap/tidb/pull/23519) for that.

We may solve it by:

1. Copy `setFromString` to tests(it is a simple helper, but I want DRY)
2. Move the tests into a non-`_tests` package(Don't want to extract these tests, )
3. Export `setFromString`

**Review by tiancaiamao — APPROVED** [review]

**Review by tangenta — APPROVED** [review]

**Comment by xhebox** [comment]
@tiancaiamao Please give me a merge. :pray:

**Pull request merged**